File: Typing.cs
using OpenQA.Selenium;
using OpenQA.Selenium.Chrome;
using OpenQA.Selenium.Support.UI;
using System;
using System.Linq;
using WindowsInput;
using System.Threading;

class Program
{
    static void Main(string[] args)
    {
        // WebDriver Chrome
        var options = new ChromeOptions();
        //options.AddArgument("--headless=new");
        // adding argument to disable the AutomationControlled flag 
        options.AddArgument("--disable-blink-features=AutomationControlled");
        // exclude the collection of enable-automation switches 
        options.AddExcludedOption("enable-automation");
        // turn-off userAutomationExtension 
        options.AddAdditionalChromeOption("useAutomationExtension", false);
        options.AddArgument("--disable-extensions");
        options.AddArgument("--profile-directory=Default");
        options.AddArgument("--incognito");
        options.AddArgument("--disable-plugins-discovery");
        options.AddArgument("--start-maximized");

        using (var driver = new ChromeDriver(options))
        {
            while (true)
            {
                driver.Navigate().GoToUrl("https://humanbenchmark.com/login");

                try
                {
                    // Wait for the element to be present (up to 10 seconds)
                    var wait = new WebDriverWait(driver, TimeSpan.FromSeconds(10));
                    var usernameField = wait.Until(d => d.FindElement(By.XPath("//input[@name='username']")));
                    
                    // Interact with the field
                    usernameField.Click();          // Focus the field
                    usernameField.Clear();          // Clear existing text (if any)
                    usernameField.SendKeys("micpowers98@gmail.com");  // Type your username
                    
                    Console.WriteLine("Successfully entered username!");
                }
                catch (Exception e)
                {
                    Console.WriteLine($"Error: {e.Message}");
                }

                try
                {
                    // Wait for the element to be present (up to 10 seconds)
                    var wait = new WebDriverWait(driver, TimeSpan.FromSeconds(10));
                    var passwordField = wait.Until(d => d.FindElement(By.XPath("//input[@name='password']")));
                    
                    // Interact with the field
                    passwordField.Click();          // Focus the field
                    passwordField.Clear();          // Clear existing text (if any)
                    passwordField.SendKeys("Idnarb10!!");  // Type your password
                    
                    Console.WriteLine("Successfully entered password!"); 
                }
                catch (Exception e)
                {
                    Console.WriteLine($"Error: {e.Message}");
                }

                try
                {
                    // Wait for the element to be present (up to 10 seconds)
                    var loginButton = driver.FindElement(By.XPath("//input[@type='submit' and @value='Login']"));
                    loginButton.Click();
                    
                    Console.WriteLine("Successfully logged in!");    
                }
                catch (Exception e)
                {
                    Console.WriteLine($"Error: {e.Message}");
                }

                Thread.Sleep(3000);

                try
                {
                    // Wait for the play link to be clickable
                    var playLink = driver.FindElement(By.CssSelector("a[href*='/tests/typing'] svg[data-icon='play-circle']")).FindElement(By.XPath(".."));
                    
                    Thread.Sleep(1000);
                    // Click the link
                    playLink.Click();
                    Console.WriteLine("Successfully clicked Play button");
                }
                catch (Exception e)
                {
                    Console.WriteLine($"Error: {e.Message}");
                }

                try
                {
                    // Wait for the letters container to load
                    var wait = new WebDriverWait(driver, TimeSpan.FromSeconds(10));
                    var lettersContainer = wait.Until(d => d.FindElement(By.ClassName("letters")));

                    // Click the typing area to focus it
                    var typingArea = driver.FindElement(By.ClassName("letters"));
                    typingArea.Click();
                    
                    // Extract all character spans
                    var characterSpans = lettersContainer.FindElements(By.TagName("span"));
                    
                    // Create a list of characters (including spaces and punctuation)
                    var characters = characterSpans.Select(span => span.Text.Length > 0 ? span.Text : " ").ToArray();
                    
                    // Type the characters using WindowsInput
                    var inputSimulator = new InputSimulator();
                    inputSimulator.Keyboard.TextEntry(string.Join("", characters));
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.Message);
                }

                bool save = false;
                try
                {
                    // Wait for the WPM element to load
                    var wait = new WebDriverWait(driver, TimeSpan.FromSeconds(10));
                    var wpmElement = wait.Until(d => d.FindElement(By.CssSelector("h1.css-0")));
                    
                    // Extract the text (e.g., "15553wpm")
                    var wpmText = wpmElement.Text;
                    
                    // Extract only the numeric part
                    var wpmNumberStr = new string(wpmText.Where(char.IsDigit).ToArray());
                    if (int.TryParse(wpmNumberStr, out int wpmNumber))
                    {
                        if (wpmNumber >= 25000)
                        {
                            save = true;
                        }
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine($"Error: {e.Message}");
                }
                
                try
                {
                    // Wait for button to be clickable (up to 10 seconds)
                    var wait = new WebDriverWait(driver, TimeSpan.FromSeconds(10));
                    var saveButton = wait.Until(d => d.FindElement(By.CssSelector("button.css-qm6rs9")));
                    
                    // Click the button
                    if (save)
                    {
                        saveButton.Click();
                        Console.WriteLine("Successfully clicked 'Save score'");
                    }
                    else
                    {
                        Console.WriteLine("Score too low.");
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine($"Error: {e.Message}");
                }
            }
        }
    }
}

File: Typing.py
from selenium.webdriver.common.by import By
from selenium.webdriver import Chrome
from selenium.webdriver.support.wait import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.remote.webelement import WebElement
from time import sleep
from warnings import filterwarnings
from Global import *
from pyautogui import typewrite,PAUSE
from random import uniform

def play(driver:Chrome,logger:Logger) -> None|Exception:
    try:
        # Wait for the letters container to load
        letters_container:WebElement = WebDriverWait(driver, 10).until(
            EC.presence_of_element_located((By.CLASS_NAME, "letters"))
        )
        logger.info("Letter container found.")
        
        # Extract all character spans
        character_spans:list[WebElement] = letters_container.find_elements(By.TAG_NAME, "span")
        logger.info("Character spans created.")
        
        # Create a list of characters (including spaces and punctuation)
        characters:list[str] = [span.text if len(span.text)>0 else " " for span in character_spans]
        logger.info("List of characters created from span.")
        message:str = "".join(characters)
        logger.info(f"Message created: {message}")

        # Click the typing area to focus it
        typing_area:WebElement = driver.find_element(By.CLASS_NAME, "letters")
        typing_area.click()
        logger.info("Typing area found and clicked")

        #typing_area.send_keys(message)
        typewrite(message,interval=uniform(0.1,0.2))
        logger.info("Typing test successfully completed.")
        return None
    except Exception as e:
        logger.critical(f"Typing test failed. Official Error: {str(e)}. Terminating program.")
        raise Exception(f"Typing test failed. Official Error: {str(e)}. Terminating program.")
    
def analyze_wpm_results(driver:Chrome,logger:Logger,min_score:int) -> bool|Exception:
    try:
        # Wait for the WPM element to load
        wpm_element:WebElement = WebDriverWait(driver, 10).until(
            EC.presence_of_element_located((By.CSS_SELECTOR, "h1.css-0"))
        )
        
        # Extract the text (e.g., "15553wpm")
        wpm_text:str = wpm_element.text
        
        # Extract only the numeric part using string manipulation
        wpm_number = ''.join(filter(str.isdigit, wpm_text))  # "15553"
        # OR (if the format is consistent)
        wpm_number = wpm_text.replace("wpm", "")  # "15553"
        if(wpm_number.isdigit()):
            wpm_number = int(wpm_number)
        save:bool = False
        if(isinstance(wpm_number,int) and wpm_number>=min_score):
            save:bool = True
        logger.info(f"Words per minute successfully analyzed. Typing speed of {wpm_number:,.0f} {'will' if save else 'will not'} be saved.")
        return save
    except Exception as e:
        logger.critical(f"Failed to analyze words per minute. Official error: {str(e)}. Terminating program.")
        raise Exception(f"Failed to analyze words per minute. Official error: {str(e)}. Terminating program.")

if __name__ == "__main__":
    filterwarnings('ignore')
    logger:Logger = create_logger()
    driver:Chrome = load_driver(logger)
    username,password,post_test_delay,\
        visual_memory,visual_memory_max_score,\
            verbal_memory,verbal_memory_max_score,\
                typing,typing_min_score,\
                    number_memory,number_memory_max_score,\
                        reaction_time,reaction_time_max_score,\
                            sequence_memory,sequence_memory_max_score,\
                                aim_trainer,chimp_test = load_configuration()
    if(not(typing)):
        logger.info("Typing in config.json set to false. Terminating program.")
    else:
        driver.get("https://humanbenchmark.com/login")

        input_username(driver,username,logger)
        input_password(driver,password,logger)
        click_login(driver,logger)  

        open_game(driver,logger,"typing")
        play(driver,logger)
        
        sleep(post_test_delay)
        save:bool = analyze_wpm_results(driver,logger,typing_min_score)
        if(save):
            save_score(driver,logger)

        driver.close()